### PR TITLE
refactor: CanisterId::try_from_principal_id.

### PR DESCRIPTION
**Pull request by daniel-wong-dfinity-org**
CanisterId::try_from(principal_id) never actually returns Err, even though not all principal IDs are canister IDs.

This new method does what CanisterId::try_from(principal_id) should have done.

# Future Work

We should look at all callers of `CanisterId::try_from(principal_id)` (including try_into) and figure out what they should call instead:

* `CanisterId::unchecked_from_principal`: This is for callers who assume that `try_from` always returns `Ok` even if the input is invalid. E.g. they immediately do `.unwrap()`, and assume that it never panics.

* `CanisterId::try_from_principal_id`: This is for callers who do NOT make that assumption. That is, they are prepared for Err to be returned. In some cases, they would even prefer that Err be returned, even though so far, `try_from` does not do that.

A simple text search would find all direct callers of `try_from`, but a simple text search would NOT find callers of `try_into` (this is (one reason) why `try_into` is "evil"). To overcome this, we can find all the code that needs to be addressed by commenting out

```
impl TryFrom<PrincipalId> for CanisterId { ...
```

and building everything. I have already made significant progress on that, but those changes should not be in this PR, because this _can_ stand alone, and already adds some value by itself, even though there are hardly any callers (so far). In fact, all those migration changes should probably not be in a single PR, because it would require approval from many teams.

After we get all tests passing with the aforementioned code commented out, we can uncomment, and change the implementation of `try_from` to simply call `try_from_principal_id`. Then, we can change all callers of `try_from_principal_id` back to `try_from`.

Unfortunately, some _other_ "secondary" `TryFrom` implementations (e.g. `impl TryFrom<&[u8]> for CanisterId`) depend on the main one (i.e. `CanisterId::try_from(principal_id)`). Callers of secondary `TryFrom` implementations should be migrated in a similar way.

### CODE REVIEW

File: rs/types/base_types/src/canister_id.rs [hunk 1]
```diff
@@ -90,6 +93,43 @@ impl CanisterId {
...
+        if raw.len() != 10 {
+            return Err(CanisterIdError::InvalidPrincipalId(format!(
+                "Principal ID {} consists of {} bytes (not 10).",
+                principal_id,
+                raw.len(),
+            )));
+        }
+
+        // Byte 8 (penultimate) must be 0x01.
+        if raw[8] != 0x01 {
```
Comment (dsarlis) [inline rs/types/base_types/src/canister_id.rs:123]: I'm not sure this is a check that can be considered conclusive. If it's opaque and the penultimate byte is 0x01 then you know it's a principal that was created by the IC and it could be a canister id, but I think it could also be a subnet id or node id. Or at the very least you have no guarantee that it's a canister id only.

I don't think it changes much for what you're trying to do here though -- we can still proceed with this as it does improve current state in the sense that if you want to handle errors or expect errors when trying to convert from a principal to a potential canister id, then this is now an actual option.

File: rs/types/base_types/src/canister_id.rs [hunk 2]
```diff
@@ -121,9 +167,9 @@ impl TryFrom<PrincipalId> for CanisterId {
 impl TryFrom<&[u8]> for CanisterId {
     type Error = CanisterIdError;
     fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
-        Self::try_from(
+        Ok(Self::unchecked_from_principal(
             PrincipalId::try_from(bytes).map_err(CanisterIdError::PrincipalIdParseError)?,
-        )
+        ))
```
Comment (dsarlis) [inline rs/types/base_types/src/canister_id.rs:172]: Is this change necessary to be in this PR? I'd expect that we actually care about this once we try the migration of `try_from` to use `try_from_principal_id` internally?

File: rs/types/base_types/src/canister_id.rs [hunk 3]
```diff
@@ -187,14 +233,12 @@ impl<'de> Deserialize<'de> for CanisterId {
...
-        // the door to invariant violation.
-        let res = CanisterId::try_from(PrincipalId::deserialize(deserializer)?);
-        let id = res.map_err(D::Error::custom)?;
-        Ok(id)
+        let result = PrincipalId::deserialize(deserializer)?;
+        // TODO: Ideally, instead of unchecked_from_principal, we'd use
+        // try_from_principal_id, but that would break people who are relying on
+        // the original behavior of CanisterId::try_from(principal_id), which
+        // would ALWAYS return Ok.
+        Ok(CanisterId::unchecked_from_principal(result))
```
Comment (dsarlis) [inline rs/types/base_types/src/canister_id.rs:241]: Similar comment as above. If we plan to change `try_from` eventually, then we might want to wait until that PR to touch this here?

**Comment by dsarlis** [comment]
Depending on how you address comments, this could then be actually a `feat` (if you constrain changes to only adding the new method and associated tests).

File: rs/types/base_types/src/canister_id.rs (hunk 2 above)
Comment (daniel-wong-dfinity-org) [reply rs/types/base_types/src/canister_id.rs]: Removed.

No, we do not need these changes in this PR (because this is a behavior preserving change).

What I was envisioning is that we'd get to the point where we can comment out `impl CanisterId from PrincipalId { ... }`, and everything would _still_ build (and all tests would still pass). Then, we can be sure that nobody is counting on the old behavior, since nobody is even calling it. Once we get to that point, then, all callers of try_from_principal_id could be converted (back) to CanisterId::try_from.

This is just an example of what you would have to do to reach "nobody is relying on the old behavior".

Yes, the goal state is for other try_from implementations to reduce to the one for PrincipalId.

Ideally, you would be able to take _any_ type that converts to PrincipalId, and be able to (directly) convert it to CanisterId. There might be a way to do this, but [when I tried], I couldn't get it to work.

[when I tried]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e023aa043771fdf1beb7af593845a383

File: rs/types/base_types/src/canister_id.rs (hunk 3 above)
Comment (daniel-wong-dfinity-org) [reply rs/types/base_types/src/canister_id.rs]: Done.

**Comment by daniel-wong-dfinity-org** [comment]
> Depending on how you address comments, this could then be actually a `feat` (if you constrain changes to only adding the new method and associated tests).

Since this change does not change any existing behaviors, I wouldn't call it a feat.

File: rs/types/base_types/src/canister_id.rs (hunk 1 above)
Comment (dsarlis) [reply rs/types/base_types/src/canister_id.rs:123]: Correct, my comment was purely informational. I agree that if we want to fix the issue that `from_u64` can return potentially non-canister IDs, it should be addressed separately.

**Comment by dsarlis** [comment]
> > Depending on how you address comments, this could then be actually a `feat` (if you constrain changes to only adding the new method and associated tests).
> 
> Since this change does not change any existing behaviors, feat is not the natural choice for me. However, I am not very against calling this a feat.

We're not changing existing behaviour but we're adding a new method to convert from a principal to canister id which will be used in the future. In a sense, this is not much different than building parts of a feature in steps, calling the individual steps "feat" but it only really changes behaviour when the feature is complete (I think we do that, at least in our team and from what I've seen in other teams). This is my line of reasoning.

That said, it was a suggestion and in no way a blocker (hence I approved).

**Review by dsarlis — APPROVED** [review]
